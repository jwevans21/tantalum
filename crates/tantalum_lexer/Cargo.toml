[package]
name = "tantalum_lexer"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}
pretty_assertions = { workspace = true }
rstest = { workspace = true }


[lints]
workspace = true
