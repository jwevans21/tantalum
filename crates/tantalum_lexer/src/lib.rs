//! # Tantalum Lexer
//!
//! This crate provides a lexer for the Tantalum language.

use tantalum_source::{SourceFile, SourceLocation, SourceSpan};
use tantalum_syntax::{SyntaxKind, SyntaxToken};

/// A lexer for the Tantalum language.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Lexer<'file> {
    /// The source file that the lexer is currently lexing.
    file: &'file SourceFile,
    /// The current location in the source file.
    location: SourceLocation,
}

impl<'file> Lexer<'file> {
    /// Create a new lexer from a source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let lexer = Lexer::new(file);
    /// ```
    #[must_use]
    pub fn new(file: &'file SourceFile) -> Self {
        Self {
            file,
            location: file.location_start(),
        }
    }

    /// Get the source file that the lexer is currently lexing.
    #[must_use]
    pub fn source(&self) -> &'file SourceFile {
        self.file
    }

    /// Get the current location in the source file.
    #[must_use]
    pub fn location(&self) -> SourceLocation {
        self.location
    }

    /// Get the character at a location in the source file.
    fn at_location(&self, location: SourceLocation) -> Option<char> {
        self.file.read_char(location)
    }

    /// Extract the next token from the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.next_token().expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    /// ```
    #[must_use]
    pub fn next_token(&mut self) -> Option<SyntaxToken> {
        let start = self.location;

        let (kind, end) = self.characters_to_token(start)?;

        self.location = end;

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Peek at the next token in the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.peek_token(1).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    ///
    /// let token = lexer.peek_token(2).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Whitespace);
    ///
    /// let token = lexer.peek_token(3).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Identifier);
    /// ```
    #[must_use]
    pub fn peek_token(&self, mut n: usize) -> Option<SyntaxToken> {
        let mut start = self.location;

        let (kind, end) = loop {
            n -= 1;
            let (kind, end) = self.characters_to_token(start)?;

            if n == 0 {
                break (kind, end);
            }

            start = end;
        };

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Convert a character to a [`SyntaxKind`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    fn characters_to_token(&self, start: SourceLocation) -> Option<(SyntaxKind, SourceLocation)> {
        let mut current = start;

        let c = self.at_location(current)?;
        current = current.next_by(c.len_utf8());

        let token = match c {
            '(' => SyntaxKind::LParen,
            ')' => SyntaxKind::RParen,
            '{' => SyntaxKind::LBrace,
            '}' => SyntaxKind::RBrace,
            '[' => SyntaxKind::LBracket,
            ']' => SyntaxKind::RBracket,
            '<' => SyntaxKind::LAngle,
            '>' => SyntaxKind::RAngle,
            ',' => SyntaxKind::Comma,
            ':' => SyntaxKind::Colon,
            ';' => SyntaxKind::Semicolon,
            '.' => SyntaxKind::Dot,
            '&' => SyntaxKind::Ampersand,
            '|' => SyntaxKind::Pipe,
            '^' => SyntaxKind::Caret,
            '+' => SyntaxKind::Plus,
            '-' => SyntaxKind::Minus,
            '*' => SyntaxKind::Star,
            '/' => SyntaxKind::Slash,
            '\\' => SyntaxKind::Backslash,
            '%' => SyntaxKind::Percent,
            '=' => SyntaxKind::Equals,
            '!' => SyntaxKind::Bang,
            '~' => SyntaxKind::Tilde,
            '?' => SyntaxKind::Question,
            '@' => SyntaxKind::At,
            '0'..='9' => SyntaxKind::Digit,
            '_' | 'a'..='z' | 'A'..='Z' => {
                while let Some(c) = self.at_location(current) {
                    match c {
                        '_' | 'a'..='z' | 'A'..='Z' | '0'..='9' => {
                            current = current.next_by(c.len_utf8());
                        }
                        _ => break,
                    }
                }

                match self.file.read_span(SourceSpan::new(
                    self.file.id(),
                    self.location.offset(),
                    current.offset(),
                ))? {
                    "fn" => SyntaxKind::Fn,
                    "let" => SyntaxKind::Let,
                    "return" => SyntaxKind::Return,
                    "if" => SyntaxKind::If,
                    "else" => SyntaxKind::Else,
                    "true" => SyntaxKind::True,
                    "false" => SyntaxKind::False,
                    _ => SyntaxKind::Identifier,
                }
            }
            c if c.is_whitespace() => SyntaxKind::Whitespace,
            _ => SyntaxKind::Other,
        };

        Some((token, current))
    }
}

#[test]
fn keywords() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "fn let return if else true false".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 7);

    assert_eq!(tokens[0].kind(), SyntaxKind::Fn);
    assert_eq!(tokens[1].kind(), SyntaxKind::Let);
    assert_eq!(tokens[2].kind(), SyntaxKind::Return);
    assert_eq!(tokens[3].kind(), SyntaxKind::If);
    assert_eq!(tokens[4].kind(), SyntaxKind::Else);
    assert_eq!(tokens[5].kind(), SyntaxKind::True);
    assert_eq!(tokens[6].kind(), SyntaxKind::False);
}

#[test]
fn identifiers() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "a A _ a1 A1 _1 _a _A __AA01".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 9);

    assert_eq!(tokens[0].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[1].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[2].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[3].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[4].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[5].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[6].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[7].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[8].kind(), SyntaxKind::Identifier);
}
