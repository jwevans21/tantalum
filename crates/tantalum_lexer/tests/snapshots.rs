use insta::assert_ron_snapshot;

use tantalum_lexer::Lexer;

macro_rules! snapshot_from_file {
    ($file_name:ident) => {
        #[test]
        fn $file_name() {
            let source = include_str!(concat!("../../examples/", stringify!($file_name), ".ta"));
            let lexer = Lexer::new(concat!(stringify!($file_name), ".ta"), source);

            let tokens = lexer.collect::<Vec<_>>();

            assert_ron_snapshot!(tokens);
        }
    };
    ($($file_name:ident),*) => {
        $(snapshot_from_file! { $file_name })*
    };
}

snapshot_from_file! {
    start,
    hello_world,
    conditionals,
    loops
}
