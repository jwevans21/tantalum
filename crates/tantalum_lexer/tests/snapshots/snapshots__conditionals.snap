---
source: tantalum_lexer/tests/snapshots.rs
expression: tokens
snapshot_kind: text
---
[
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 0,
        line: 1,
        column: 1,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 2,
        line: 1,
        column: 3,
      ),
    ),
    data: Token(
      lexeme: "fn",
      kind: KeywordFn,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 3,
        line: 1,
        column: 4,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 7,
        line: 1,
        column: 8,
      ),
    ),
    data: Token(
      lexeme: "main",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 7,
        line: 1,
        column: 8,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 8,
        line: 1,
        column: 9,
      ),
    ),
    data: Token(
      lexeme: "(",
      kind: LeftParen,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 8,
        line: 1,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 9,
        line: 1,
        column: 10,
      ),
    ),
    data: Token(
      lexeme: ")",
      kind: RightParen,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 9,
        line: 1,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 10,
        line: 1,
        column: 11,
      ),
    ),
    data: Token(
      lexeme: ":",
      kind: Colon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 11,
        line: 1,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 14,
        line: 1,
        column: 15,
      ),
    ),
    data: Token(
      lexeme: "i32",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 15,
        line: 1,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 16,
        line: 1,
        column: 17,
      ),
    ),
    data: Token(
      lexeme: "{",
      kind: LeftBrace,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 21,
        line: 2,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 24,
        line: 2,
        column: 8,
      ),
    ),
    data: Token(
      lexeme: "let",
      kind: KeywordLet,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 25,
        line: 2,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 26,
        line: 2,
        column: 10,
      ),
    ),
    data: Token(
      lexeme: "a",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 26,
        line: 2,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 27,
        line: 2,
        column: 11,
      ),
    ),
    data: Token(
      lexeme: ":",
      kind: Colon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 28,
        line: 2,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 31,
        line: 2,
        column: 15,
      ),
    ),
    data: Token(
      lexeme: "i32",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 32,
        line: 2,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 33,
        line: 2,
        column: 17,
      ),
    ),
    data: Token(
      lexeme: "=",
      kind: Equal,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 34,
        line: 2,
        column: 18,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 36,
        line: 2,
        column: 20,
      ),
    ),
    data: Token(
      lexeme: "10",
      kind: DecimalIntegerLiteral,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 36,
        line: 2,
        column: 20,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 37,
        line: 2,
        column: 21,
      ),
    ),
    data: Token(
      lexeme: ";",
      kind: Semicolon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 42,
        line: 3,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 45,
        line: 3,
        column: 8,
      ),
    ),
    data: Token(
      lexeme: "let",
      kind: KeywordLet,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 46,
        line: 3,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 47,
        line: 3,
        column: 10,
      ),
    ),
    data: Token(
      lexeme: "b",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 47,
        line: 3,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 48,
        line: 3,
        column: 11,
      ),
    ),
    data: Token(
      lexeme: ":",
      kind: Colon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 49,
        line: 3,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 52,
        line: 3,
        column: 15,
      ),
    ),
    data: Token(
      lexeme: "i32",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 53,
        line: 3,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 54,
        line: 3,
        column: 17,
      ),
    ),
    data: Token(
      lexeme: "=",
      kind: Equal,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 55,
        line: 3,
        column: 18,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 57,
        line: 3,
        column: 20,
      ),
    ),
    data: Token(
      lexeme: "20",
      kind: DecimalIntegerLiteral,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 57,
        line: 3,
        column: 20,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 58,
        line: 3,
        column: 21,
      ),
    ),
    data: Token(
      lexeme: ";",
      kind: Semicolon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 63,
        line: 4,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 66,
        line: 4,
        column: 8,
      ),
    ),
    data: Token(
      lexeme: "let",
      kind: KeywordLet,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 67,
        line: 4,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 68,
        line: 4,
        column: 10,
      ),
    ),
    data: Token(
      lexeme: "c",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 68,
        line: 4,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 69,
        line: 4,
        column: 11,
      ),
    ),
    data: Token(
      lexeme: ":",
      kind: Colon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 70,
        line: 4,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 73,
        line: 4,
        column: 15,
      ),
    ),
    data: Token(
      lexeme: "i32",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 74,
        line: 4,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 75,
        line: 4,
        column: 17,
      ),
    ),
    data: Token(
      lexeme: "=",
      kind: Equal,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 76,
        line: 4,
        column: 18,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 78,
        line: 4,
        column: 20,
      ),
    ),
    data: Token(
      lexeme: "30",
      kind: DecimalIntegerLiteral,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 78,
        line: 4,
        column: 20,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 79,
        line: 4,
        column: 21,
      ),
    ),
    data: Token(
      lexeme: ";",
      kind: Semicolon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 85,
        line: 6,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 87,
        line: 6,
        column: 7,
      ),
    ),
    data: Token(
      lexeme: "if",
      kind: KeywordIf,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 88,
        line: 6,
        column: 8,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 89,
        line: 6,
        column: 9,
      ),
    ),
    data: Token(
      lexeme: "a",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 90,
        line: 6,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 91,
        line: 6,
        column: 11,
      ),
    ),
    data: Token(
      lexeme: ">",
      kind: RightAngle,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 92,
        line: 6,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 93,
        line: 6,
        column: 13,
      ),
    ),
    data: Token(
      lexeme: "b",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 94,
        line: 6,
        column: 14,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 95,
        line: 6,
        column: 15,
      ),
    ),
    data: Token(
      lexeme: "{",
      kind: LeftBrace,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 104,
        line: 7,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 110,
        line: 7,
        column: 15,
      ),
    ),
    data: Token(
      lexeme: "return",
      kind: KeywordReturn,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 111,
        line: 7,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 112,
        line: 7,
        column: 17,
      ),
    ),
    data: Token(
      lexeme: "a",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 112,
        line: 7,
        column: 17,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 113,
        line: 7,
        column: 18,
      ),
    ),
    data: Token(
      lexeme: ";",
      kind: Semicolon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 118,
        line: 8,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 119,
        line: 8,
        column: 6,
      ),
    ),
    data: Token(
      lexeme: "}",
      kind: RightBrace,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 120,
        line: 8,
        column: 7,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 124,
        line: 8,
        column: 11,
      ),
    ),
    data: Token(
      lexeme: "else",
      kind: KeywordElse,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 125,
        line: 8,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 127,
        line: 8,
        column: 14,
      ),
    ),
    data: Token(
      lexeme: "if",
      kind: KeywordIf,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 128,
        line: 8,
        column: 15,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 129,
        line: 8,
        column: 16,
      ),
    ),
    data: Token(
      lexeme: "b",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 130,
        line: 8,
        column: 17,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 131,
        line: 8,
        column: 18,
      ),
    ),
    data: Token(
      lexeme: ">",
      kind: RightAngle,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 132,
        line: 8,
        column: 19,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 133,
        line: 8,
        column: 20,
      ),
    ),
    data: Token(
      lexeme: "c",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 134,
        line: 8,
        column: 21,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 135,
        line: 8,
        column: 22,
      ),
    ),
    data: Token(
      lexeme: "{",
      kind: LeftBrace,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 144,
        line: 9,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 150,
        line: 9,
        column: 15,
      ),
    ),
    data: Token(
      lexeme: "return",
      kind: KeywordReturn,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 151,
        line: 9,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 152,
        line: 9,
        column: 17,
      ),
    ),
    data: Token(
      lexeme: "b",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 152,
        line: 9,
        column: 17,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 153,
        line: 9,
        column: 18,
      ),
    ),
    data: Token(
      lexeme: ";",
      kind: Semicolon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 158,
        line: 10,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 159,
        line: 10,
        column: 6,
      ),
    ),
    data: Token(
      lexeme: "}",
      kind: RightBrace,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 160,
        line: 10,
        column: 7,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 164,
        line: 10,
        column: 11,
      ),
    ),
    data: Token(
      lexeme: "else",
      kind: KeywordElse,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 165,
        line: 10,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 166,
        line: 10,
        column: 13,
      ),
    ),
    data: Token(
      lexeme: "{",
      kind: LeftBrace,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 175,
        line: 11,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 181,
        line: 11,
        column: 15,
      ),
    ),
    data: Token(
      lexeme: "return",
      kind: KeywordReturn,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 182,
        line: 11,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 183,
        line: 11,
        column: 17,
      ),
    ),
    data: Token(
      lexeme: "c",
      kind: Identifier,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 183,
        line: 11,
        column: 17,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 184,
        line: 11,
        column: 18,
      ),
    ),
    data: Token(
      lexeme: ";",
      kind: Semicolon,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 189,
        line: 12,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 190,
        line: 12,
        column: 6,
      ),
    ),
    data: Token(
      lexeme: "}",
      kind: RightBrace,
    ),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 191,
        line: 13,
        column: 1,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 192,
        line: 13,
        column: 2,
      ),
    ),
    data: Token(
      lexeme: "}",
      kind: RightBrace,
    ),
  ),
]
